### PR TITLE
Read line handling

### PR DESCRIPTION
**Pull request by brucemiller**
Subtle & Fiddly: careful handling of lines by `\read` and handling of `\endlinechar`. This allows `\read` to be simplified.
The io test case now covers more corner cases in the data, as well as `\endlinechar`.  I *think* I've got it right, now!

Along with a solution to #1511, this should handle #1505.

@xworld21 You might want to take a peek at this, since it reworks your previous patch #1469.

### CODE REVIEW

**Comment by dginev** [comment]
The code certainly looks quite subtle. I think I'll start by reading through the involved test file PDFs and XMLs and convincing myself they look identical - and if that's the case, might as well merge things right away.

File: lib/LaTeXML/Core/Mouth.pm [hunk 1]
```diff
@@ -288,49 +284,53 @@ sub readToken {
...
       $$self{colno} = 0;
       my $line = $self->getNextLine;
+      # For \read, we have to return something for EOL, and handle implicit final newline
+      my $read_mode = (($STATE->lookupValue('PRESERVE_NEWLINES') || 0) > 1);
+      my $eolch     = "\r";
+      if (my $eol = $STATE->lookupDefinition(T_CS('\endlinechar'))) {
+        $eol   = $eol->valueOf()->valueOf;
+        $eolch = (($eol > 0) && ($eol <= 255) ? chr($eol) : undef); }
       if (!defined $line) {    # Exhausted the input.
+        my $eolcc    = ((defined $eolch) && $STATE->lookupCatcode($eolch)) // CC_OTHER;
```
Comment (xworld21) [inline lib/LaTeXML/Core/Mouth.pm:294]: @brucemiller a bit tangential, but it would be more robust to have `lookupCatcode` return `CC_OTHER` by default, like TeX does. Right now you add `// CC_OTHER` to every call to `lookupCode`, I think this would become the fifth line, see e.g. [this patch](https://github.com/brucemiller/LaTeXML/compare/master...xworld21:b66a301).

**Comment by brucemiller** [comment]
@xworld21; yeah, that would probably be a safer thing to do, and avoid 
some (cough) occasional hickups.

The original expectation was that I might want to know if it had been 
assigned or not, so return undef. But that doesn't seem to have been 
needed; it never got reexamined, till now :>

**Review by xworld21 — COMMENTED** [review]
Oh well, I have tried doing a code review. As far as I can tell, it is orthogonal to my PR which only affected normal (and `\input`) parsing, and in fact it passes my additional test, so that's good.

Just two comments: can the `$read_mode = 1` behaviour be implemented in `getNextLine`? I figure you can get the same result with fewer special purpose code (I could be very wrong, of course). Also, I didn't understand the `T_MARKER` bit, but then I don't know how markers are supposed to behave.

File: lib/LaTeXML/Core/Mouth.pm [hunk 2]
```diff
@@ -288,49 +284,53 @@ sub readToken {
...
+
       $$self{chars}  = splitChars($line);
       $$self{nchars} = scalar(@{ $$self{chars} });
       # In state N, skip spaces
       while (($$self{colno} < $$self{nchars})
         # DIRECT ACCESS to $STATE's catcode table!!!
         && (($$STATE{catcode}{ $$self{chars}[$$self{colno}] }[0] || CC_OTHER) == CC_SPACE)) {
         $$self{colno}++; }
-
+      return T_MARKER('EOL') if $read_mode && ((!defined $eolch) || ($eolch ne "\r"));
```
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm:315]: Well, that really is the end of the line, but it returns the end-line Marker *after* it has prepared the next line (if any) for input.

File: lib/LaTeXML/Core/Mouth.pm [hunk 3]
```diff
@@ -288,49 +284,53 @@ sub readToken {
...
       my $line = $self->getNextLine;
+      # For \read, we have to return something for EOL, and handle implicit final newline
+      my $read_mode = (($STATE->lookupValue('PRESERVE_NEWLINES') || 0) > 1);
+      my $eolch     = "\r";
+      if (my $eol = $STATE->lookupDefinition(T_CS('\endlinechar'))) {
+        $eol   = $eol->valueOf()->valueOf;
+        $eolch = (($eol > 0) && ($eol <= 255) ? chr($eol) : undef); }
       if (!defined $line) {    # Exhausted the input.
+        my $eolcc    = ((defined $eolch) && $STATE->lookupCatcode($eolch)) // CC_OTHER;
+        my $eoftoken = $read_mode && (defined $eolch) && !$$self{at_eof} && $$self{source}
```
[the remark this answers is not in the document]
Comment (dginev) [reply lib/LaTeXML/Core/Mouth.pm:295]: "Better safe than sorry" is a pretty good strategy in Perl, might as well keep both :>

File: lib/LaTeXML/Core/Mouth.pm [hunk 4]
```diff
@@ -288,49 +284,53 @@ sub readToken {
       $$self{lineno}++;
       $$self{colno} = 0;
       my $line = $self->getNextLine;
+      # For \read, we have to return something for EOL, and handle implicit final newline
+      my $read_mode = (($STATE->lookupValue('PRESERVE_NEWLINES') || 0) > 1);
+      my $eolch     = "\r";
+      if (my $eol = $STATE->lookupDefinition(T_CS('\endlinechar'))) {
+        $eol   = $eol->valueOf()->valueOf;
+        $eolch = (($eol > 0) && ($eol <= 255) ? chr($eol) : undef); }
```
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm:292]: Possibly, but would still require a careful testing to avoid a Perl trap. I used `\enlinechar` being "0" to force myself to get it (partly) right, since "" and "0" are both "false" from Perl's perspective.

File: lib/LaTeXML/Core/Mouth.pm [hunk 5]
```diff
@@ -288,49 +284,53 @@ sub readToken {
...
       if (!defined $line) {    # Exhausted the input.
+        my $eolcc    = ((defined $eolch) && $STATE->lookupCatcode($eolch)) // CC_OTHER;
+        my $eoftoken = $read_mode && (defined $eolch) && !$$self{at_eof} && $$self{source}
+          && ($eolcc == CC_EOL ? T_CS('\par')
+          : Token($eolch, $eolcc));
```
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm:297]: If I understand, would just duplicate different code, since `getNextLine` is subclassed in `LaTeXML::Mouth::file`.

File: lib/LaTeXML/Core/Mouth.pm (hunk 3 above)
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm:295]: Testing eof isn't always done, usually it's reading until you get an undef Token; so w/o some corresponding tweaks elsewhere, this gives an infinite loop.  (with the current patch, the eof test may no longer be needed in `\read`)

It certainly can be that I haven't quite captured TeX's internal logic, but many of these Token reading effects don't take effect until the *next* read (skipping spaces, for example).

File: lib/LaTeXML/Core/Mouth.pm [hunk 6]
```diff
@@ -224,7 +224,7 @@ sub handle_space {
   my ($ch, $cc);
   # Skip any following spaces!
   while ((($ch, $cc) = getNextChar($self)) && (defined $ch) && (($cc == CC_SPACE) || ($cc == CC_EOL))) { }
-  $$self{colno}-- if ($$self{colno} < $$self{nchars});
+  $$self{colno}-- if ($$self{colno} <= $$self{nchars}) && (defined $ch);    # backup at nonspace/eol
```
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm:227]: There were times when I thought setting `colno` *past* `nchars` would be a good idea, but I don't think that happens now. But I think I'd second Deyan's comment :>

**Comment by brucemiller** [comment]
General comment on `T_MARKER`. It's a LaTeXML extension, not part of TeX.  Such tokens should never escape out into the actual processing machinery, but simply marks a place in the token stream for some obscure purpose. Like profiling macro execution or, in this case, being able to return *something* to mark the end of line, when otherwise no token would be created.

**Review by dginev — APPROVED** [review]
Ok, reviewing the test PDF+XML combination wasn't that bad at all, quite systematic mostly.

And everything looks as if it matches nicely - so it's definitely an upgrade to the master branch. LGTM!

File: lib/LaTeXML/Core/Mouth.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (xworld21) [reply lib/LaTeXML/Core/Mouth.pm:315]: Eheh, I still don't get it. When you run this on the very first line, don't you get a marker *before* all the other tokens?

**Comment by xworld21** [comment]
Thanks for clarifying!

File: lib/LaTeXML/Core/Mouth.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (dginev) [reply lib/LaTeXML/Core/Mouth.pm:315]: @xworld21 if I am getting your question correctly, the answer is to notice when line 315 is actually entered - which is when the condition on line 283 is true:
```
 if ($$self{colno} >= $$self{nchars}) {
```

So this T_MARKER will get created only in the circumstance where one line has been completed, and the next line has been setup, i.e. at the exact time the Mouth transitions to reading the next line. So you're right this will happen *before* all tokens, but those are the *next* tokens, after the line break.

**Comment by brucemiller** [comment]
Except that that clause is used to setup the 1st line, too, so @xworld21 
is correct that a marker is returned before the 1st line. But it has no 
effect, since it gets ignored and `isEOL` returns false.

But more annoying (although it has no visible effect) is that markers 
are returned even after non-empty lines (which would have triggered 
isEOL).  The marker is really only needed when an empty line is read and 
there is no otherwise recognizable EOL token to return.

Thanks; Patch en route!

File: lib/LaTeXML/Core/Mouth.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm]: Thanks for that! The test is slightly clunkier, but only returns the marker exactly when needed.

File: lib/LaTeXML/Core/Mouth.pm [hunk 7]
```diff
@@ -281,6 +281,8 @@ sub readToken {
   while (1) {    # Iterate till we find a token, or run out. (use return)
                  # ===== Get next line, if we need to.
     if ($$self{colno} >= $$self{nchars}) {
+      my $lastn   = $$self{nchars};
+      my $lastcol = $$self{colno};
```
Comment (dginev) [inline lib/LaTeXML/Core/Mouth.pm:285]: `$lastn` and `$lastcol` are never used? Is there a missing commit?

Comment (brucemiller) [reply lib/LaTeXML/Core/Mouth.pm]: nope, missing cleanup. Ugh.

File: lib/LaTeXML/Core/Mouth.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (xworld21) [reply lib/LaTeXML/Core/Mouth.pm]: Ah, now I see what you were trying to do! It all makes sense now.

**Review by xworld21 — APPROVED** [review]
Not that you need my approval, but I have to close the review in some way!
